# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: modules/libjar/nsJAR.cpp
# Commit: a345e05df151
# Full Hash: a345e05df151563625fe05ec63094b3dae2c9379
# Author: Doug Thayer <dothayer@mozilla.com>
# Date: 2020-07-07 21:31:11
# Regressor Bug: 1627075
# File Overlap Count: 1
# Description:
#   Bug 1627075 - Route Omnijar requests through StartupCache r=froydnj
#   
#   This should be a relatively straightforward patch. Essentially, we implement
#   a wrapper class (and friends) around nsZipArchive (and friends), which transparently
#   caches entries from the underlying zip archive in the StartupCache. This will break
# ==============================================================================

diff -r 288a67aed661 -r a345e05df151 modules/libjar/nsJAR.cpp
--- a/modules/libjar/nsJAR.cpp	Tue Jul 07 04:35:08 2020 +0000
+++ b/modules/libjar/nsJAR.cpp	Tue Jul 07 04:36:27 2020 +0000
@@ -26,7 +26,7 @@
 
 // The following initialization makes a guess of 10 entries per jarfile.
 nsJAR::nsJAR()
-    : mZip(new nsZipArchive()),
+    : mZip(nullptr),
       mReleaseTime(PR_INTERVAL_NO_TIMEOUT),
       mCache(nullptr),
       mLock("nsJAR::mLock"),
@@ -87,12 +87,13 @@
 
   // The omnijar is special, it is opened early on and closed late
   // this avoids reopening it
-  RefPtr<nsZipArchive> zip = mozilla::Omnijar::GetReader(zipFile);
+  RefPtr<CacheAwareZipReader> zip = mozilla::Omnijar::GetReader(zipFile);
   if (zip) {
     mZip = zip;
     mSkipArchiveClosing = true;
     return NS_OK;
   }
+  mZip = new CacheAwareZipReader();
   return mZip->OpenArchive(zipFile);
 }
 
@@ -102,7 +103,7 @@
   if (mOpened) return NS_ERROR_FAILURE;  // Already open!
 
   nsJAR* outerJAR = static_cast<nsJAR*>(aZipReader);
-  RefPtr<nsZipArchive> innerZip =
+  RefPtr<CacheAwareZipReader> innerZip =
       mozilla::Omnijar::GetInnerReader(outerJAR->mZipFile, aZipEntry);
   if (innerZip) {
     mOpened = true;
@@ -124,11 +125,12 @@
   mOuterZipEntry.Assign(aZipEntry);
 
   RefPtr<nsZipHandle> handle;
-  rv = nsZipHandle::Init(static_cast<nsJAR*>(aZipReader)->mZip.get(),
+  rv = nsZipHandle::Init(static_cast<nsJAR*>(aZipReader)->mZip->GetZipArchive(),
                          PromiseFlatCString(aZipEntry).get(),
                          getter_AddRefs(handle));
   if (NS_FAILED(rv)) return rv;
 
+  mZip = new CacheAwareZipReader();
   return mZip->OpenArchive(handle);
 }
 
@@ -144,6 +146,7 @@
                                   getter_AddRefs(handle));
   if (NS_FAILED(rv)) return rv;
 
+  mZip = new CacheAwareZipReader();
   return mZip->OpenArchive(handle);
 }
 
@@ -165,7 +168,7 @@
   if (mSkipArchiveClosing) {
     // Reset state, but don't close the omnijar because we did not open it.
     mSkipArchiveClosing = false;
-    mZip = new nsZipArchive();
+    mZip = nullptr;
     return NS_OK;
   }
 
@@ -301,7 +304,7 @@
     return NS_ERROR_FAILURE;
   }
 
-  RefPtr<nsZipHandle> handle = mZip->GetFD();
+  RefPtr<nsZipHandle> handle = mZip->GetZipArchive()->GetFD();
   if (!handle) {
     return NS_ERROR_FAILURE;
   }